src/mds/MDSUDPEndpoint.cc

```diff
@@ -0,0 +1,92 @@
+#include "MDSUDPEndpoint.h"
+#define dout_subsys ceph_subsys_mds
+
+int MDSUDPManager::send(
+    const std::shared_ptr<MDSNotificationMessage> &message) {
+  std::shared_lock<std::shared_mutex> lock(endpoint_mutex);
+  std::vector<boost::asio::const_buffer> buf(2);
+  for (auto &[key, endpoint] : endpoints) {
+    uint64_t len = message->message.length();
+    buf[0] = boost::asio::buffer(&len, sizeof(len));
+    buf[1] = boost::asio::buffer(message->message.c_str(),
```

Comment:
Generally bufferlist::c_str() is a pretty costly opertation - it makes a contiguous buffer from a partitioned one. Which implies buffer allocation and data copying
You might want to use bufferlist API to access the parts and put them as multiple elements of the target vector.
Not urgent but it's worth some TODO remark here IMO.